fs/zbd_zenfs.h

```diff
@@ -90,32 +90,46 @@ class Zone {
   void CloseWR(); /* Done writing */
 };
 
+// Abstract class as interface.
+// operator() must be overrided in order to execute the function.
 class BackgroundJob {
  public:
-  BackgroundJob(std::function<int(void *)> fn, void *arg)
-       : fn_(fn), arg_(arg) {}
-   std::function<int(void *)> fn_;
-  void *arg_;
-  virtual void operator()() { fn_(arg_); }
+  virtual void operator()() = 0;
   virtual ~BackgroundJob() {}
 };
 
-class ErrorHandlingBGJob : public BackgroundJob {
+// For simple jobs that needs no return value nor argument.
+class SimpleJob : public BackgroundJob {
+ public:
+  std::function<void()> fn_;
+  // No default allowed.
+  SimpleJob() = delete;
+  SimpleJob(std::function<void()> fn) : fn_(fn) {}
+  virtual void operator()() override { fn_(); }
+  virtual ~SimpleJob() {}
+};
+
+template<typename arg_t, typename ret_t>
+class GeneralJob : public BackgroundJob {
  public:
-  ErrorHandlingBGJob(std::function<int(void *)> fn, void *arg,
-                      std::function<void(int)> handler)
-       : BackgroundJob(fn, arg), handler_(handler) {}
-   ErrorHandlingBGJob(std::function<int(void *)> fn, void *arg,
-                      std::function<void(int)> &&handler)
-       : BackgroundJob(fn, arg), handler_(handler) {}
-   std::function<void(int)> handler_;
-   virtual void operator()() override { handler_(fn_(arg_)); }
+  // Job requires argument and return value for error handling
+  std::function<ret_t(arg_t)> fn_;
+  // Argument for execution
+  arg_t arg_;
+  // Error handler
+  std::function<void(ret_t)> hdl_;
+  // No default allowed
+  GeneralJob() = delete;
+  GeneralJob(std::function<ret_t(arg_t)> fn, arg_t arg,
+             std::function<void(ret_t)> hdl) : fn_(fn), arg_(arg), hdl_(hdl) {}
+  virtual void operator()() override { hdl_(fn_(arg_)); }
+  virtual ~GeneralJob() {}
 };
 
 class BackgroundWorker {
   enum WorkingState { kWaiting = 0, kRunning, kTerminated } state_;
   std::thread worker_;
-  std::list<BackgroundJob> jobs_;
+  std::list<std::unique_ptr<BackgroundJob>> jobs_;
```

Comment:
Is there any strong reason that we have to use `unique_ptr` here?

Reply:
Jobs share nothing. It's safer and easier to debug when the memory structure exists.